### PR TITLE
Feature/persistent scoreboard

### PR DESCRIPTION
**Pull request by arbuckle**
This is a request for comments. /!\ DO NOT MERGE /!\

@congratulatio @wight-airmash 

I've implemented a scoreboard recovery mechanism for FFA that ensures player stats are preserved if the server is restarted.  It piggybacks on the mechanism that was put in place for CTF, but adds a dimension of persistence to server restarts. This is useful for server maintenance in FFA - the server can be restarted without upsetting players. This also preserves player stats for up to five minutes after they have left the game, eliminating a source of frustration when a player connection resets.

A few key things to note:

- Data is persisted to disk every minute in recovery.js. A synchronous write is issued, but the function is triggered through the event framework so I wouldn't expect it to block the game loop. 
- There's a notable change to shutdown behavior in bootstrap.ts.  I was unable to piggyback on the existing 'finalizers' in log.processFinalHandlers and I couldn't find another option to trigger a behavior on shutdown, so I added a two-way event mechanism.  Shutdown activities trigger a TIMELINE_SERVER_SHUTDOWN event, and the recovery module returns a TIMELINE_RECOVERY_COMPLETE event.
- I'm not entirely happy with the above. Shutdown events can trigger multiple times and there's no good way to 'register' a new shutdown handler. 
- One decision I'd like feedback on is whether persisting data to disk should be a "default" behavior that the admin can choose to disable through a /server command, or whether persistence should be triggered by a server command by the admin. I haven't implemented these commands yet.
- Another is w/r/t the game balance - does persisting stats on disconnect under circumstances other than a server restart introduce too much of a risk of exploitation by players?

### CODE REVIEW

**Comment by wight-airmash** [comment]
> Data is persisted to disk every minute in recovery.js. A synchronous write is issued, but the function is triggered through the event framework so I wouldn't expect it to block the game loop.

[Events | Node.js](https://nodejs.org/docs/latest-v12.x/api/events.html#events_events): "When the EventEmitter object emits an event, all of the functions attached to that specific event are called synchronously. Any values returned by the called listeners are ignored and will be discarded."

> Another is w/r/t the game balance - does persisting stats on disconnect under circumstances other than a server restart introduce too much of a risk of exploitation by players?

The main reason why recovery was only available in CTF is because of the disconnection abuse. It's just the perfect gift for "runners". Even in CTF (where bounty is not so important) some individuals tried to troll people using it, in FFA it is inevitable.

To prevent this, the EU CTF is implemented with a disconnection delay: player's aircraft disappears a few seconds later after the actual disconnection which gives the opponent the opportunity to kill the abuser.

**Comment by congratulatio** [comment]
I like this. I'm not sure if I'd use it for server upgrades on the EU FFA (as usually I will start the second server separately on another port, and run a script to load a new nginx config pointing to it when there's zero human players) but definitely sounds useful for this:

> This also preserves player stats for up to five minutes after they have left the game, eliminating a source of frustration when a player connection resets.

With a change to the frontend to automatically attempt a reconnect, this would be nearly seamless from the player point of view. Could even survive an IP address change if the frontend supplied some sort of session id in the `LOGIN` packet.

> Data is persisted to disk every minute in recovery.js. A synchronous write is issued, but the function is triggered through the event framework so I wouldn't expect it to block the game loop.

What are your thoughts on integrating with the code in [src/server/periodic/user-stats](https://github.com/wight-airmash/ab-server/tree/v6.3.1/src/server/periodic/user-stats) that's currently used for saving user-stats.json and sync-state.json? It uses a worker for serializing and an asynchronous write. But it doesn't do anything on shutdown, so there is currently a possibility of a few seconds data loss, which your work on the shutdown event I think would mitigate.

> One decision I'd like feedback on is whether persisting data to disk should be a "default" behavior that the admin can choose to disable through a /server command, or whether persistence should be triggered by a server command by the admin. I haven't implemented these commands yet.

I'm not sure what the default should be, but I think having it as configurable in [src/config/config.ts](https://github.com/wight-airmash/ab-server/blob/v6.3.1/src/config/config.ts) would make sense, to be consistent with the other data persistence features.

File: src/core/bootstrap.ts [hunk 1]
```diff
@@ -384,33 +390,34 @@ export default class GameServerBootstrap implements GameServerBootstrapInterface
...
   private bindExitListeners(): void {
+
     process.on('beforeExit', () => {
       this.log.processfinalHandlers(null, 'beforeExit');
     });
 
     process.on('exit', () => {
       this.log.processfinalHandlers(null, 'exit');
-      this.stop();
+      this.events.emit(TIMELINE_SERVER_SHUTDOWN)
```
Comment (congratulatio) [inline src/core/bootstrap.ts:400]: Maybe something like this to mitigate the multiple shutdown events?

```suggestion
      if (!this.shutdownInProgress) {
        this.shutdownInProgress = true;
        this.events.emit(TIMELINE_SERVER_SHUTDOWN);
      }
```

**Comment by arbuckle** [comment]
> What are your thoughts on integrating with the code in src/server/periodic/user-stats that's currently used for saving user-stats.json and sync-state.json? It uses a worker for serializing and an asynchronous write. But it doesn't do anything on shutdown, so there is currently a possibility of a few seconds data loss, which your work on the shutdown event I think would mitigate.

https://github.com/wight-airmash/ab-server/compare/develop...arbuckle:feature/persistent-scoreboard-in-user-stats

So I gave this a try and I think it's a case of moving complexity around, rather than a case of reducing it. It was fairly straightforward to extend user-stats to handle the additional use case and it even seemed to complete the write on exit (though I believe it's a race). I wasn't clear on how to handle the cache invalidation and didn't follow through on moving the restore-from-disk operation.  As an alternative I added a conditional async write into the current branch.  

I think this is the essence of it:
- Put all of the recovery handling in `recovering.ts` - this is optimizing for having the recovery procedures self-contained in a single module. 
- Move the recovery handling to `user-stats.ts` - this is optimizing for re-using a single method for writing state files to disk. 

If anyone has a strong opinion one way or the other I'd oblige.

**Comment by wight-airmash** [comment]
Nice trick with connections clearing delay, but there are some peculiarities:

- A player can create a duplicate of the entity. For example: player 1 is chasing player 2. Player 2 reconnects (as fast as possible) and safely respawns elsewhere, recovering the stats (score, upgrades). At this point the old duplicate is still in game. Player 1 kills the duplicate and gains its points, but player 2 is not actually affected.
- Creating duplicates is a potential opportunity for abuse (you can fill the server with fakes).
- Now checks such as `this.storage.connectionList.has(connectionId)` are meaningless.

Suggestions:

- When reconnecting, the player should recovering into its existed entity, but not the new one.
- Use a status flag like `player.connected`. This will require writting more logic in different places in the code, but will get rid of inconsistencies. In this case, I would suggest moving the delayed disconnect logic to the `GamePlayersDisconnect` system and keep clear `Connections` system from in-game logic.

> So I gave this a try and I think it's a case of moving complexity around, rather than a case of reducing it. It was fairly straightforward to extend user-stats to handle the additional use case and it even seemed to complete the write on exit (though I believe it's a race).

I guess the main reason @congratulatio suggested moving this to user-stats is because of the worker thread. Serialization is not an asynchronous operation, being in the game loop thread, it blocks it.

File: src/server/connections.ts [hunk 2]
```diff
@@ -152,33 +153,59 @@ export default class Connections extends System {
...
+        player.keystate.LEFT = false
+        player.keystate.RIGHT = false
+        player.keystate.SPECIAL = false
+        player.keystate.FIRE = false
+        player.shield.current = false
+        player.delayed.BROADCAST_PLAYER_UPDATE = true
+      }
+
+      let ctx = this;
+      setTimeout(function() {
```
Comment (wight-airmash) [inline src/server/connections.ts:175]: Using the arrow function here (as elsewhere in the project) will avoid creating a closure.

**Comment by congratulatio** [comment]
> I guess the main reason @congratulatio suggested moving this to user-stats is because of the worker thread. Serialization is not an asynchronous operation, being in the game loop thread, it blocks it.

I did but I am wondering about this now, as from testing this there is still the overhead from node serializing the data internally to send to the worker thread.

Maybe my test case is not properly representative though, not sure: https://gist.github.com/congratulatio/f4186a117d7416be0bc2f547f29587e6

**Comment by wight-airmash** [comment]
> I did but I am wondering about this now, as from testing this there is still the overhead from node serializing the data internally to send to the worker thread.

Yes, under the hood the data transfered with structured clone algorithm and sending ([serialization](https://nodejs.org/docs/latest-v12.x/api/v8.html#v8_v8_serialize_value)) blocks the main thread and receiving ([deserialization](https://nodejs.org/docs/latest-v12.x/api/v8.html#v8_v8_deserialize_buffer)) blocks the worker. When the data is large and often needs to be transferred, it may be more efficient to synchronize it using more memory (depending on the requirements of the application of course).

**Comment by fire-airmash** [comment]
We have a hacker in aimash names computer wise/xxx he has leaked one of the ISE codes and played his ways into the airmash system. keep accounts safe. use alts. if needed. he will be placed on hold.

**Comment by fire-airmash** [comment]
trolled lol. i'm xxx/ lol Do you like my mod i made? https://us.airmash.online/btr/admin

**Pull request closed without merging**